core/stateful_precompile_test.go

```diff
@@ -0,0 +1,252 @@
+			precompileAddr: precompile.ContractDeployerAllowListAddress,
+			input: func() []byte {
+				input, err := precompile.PackModifyAllowList(adminAddr, precompile.AllowListNoRole)
+				if err != nil {
+					panic(err)
+				}
+				return input
+			},
+			suppliedGas: precompile.ModifyAllowListGasCost,
+			readOnly:    true,
+			setupState: func(state *state.StateDB) {
+				precompile.SetContractDeployerAllowListStatus(state, adminAddr, precompile.AllowListAdmin)
+			},
+			expectedErr: precompile.ErrReadOnlyModifyAllowList.Error(),
+		},
+		"read allow list no role": {
+			caller:         adminAddr,
+			precompileAddr: precompile.ContractDeployerAllowListAddress,
+			input: func() []byte {
+				return precompile.PackReadAllowList(noRoleAddr)
+			},
+			suppliedGas: precompile.ModifyAllowListGasCost,
+			readOnly:    false,
+			setupState:  func(state *state.StateDB) {},
+			expectedRes: common.Hash(precompile.AllowListNoRole).Bytes(),
+			assertState: func(t *testing.T, state *state.StateDB) {
+				res := precompile.GetContractDeployerAllowListStatus(state, adminAddr)
+				assert.Equal(t, precompile.AllowListNoRole, res)
+			},
+		},
+		"read allow list admin role": {
+			caller:         adminAddr,
+			precompileAddr: precompile.ContractDeployerAllowListAddress,
+			input: func() []byte {
+				return precompile.PackReadAllowList(noRoleAddr)
+			},
+			suppliedGas: precompile.ModifyAllowListGasCost,
+			readOnly:    false,
+			setupState: func(state *state.StateDB) {
+				precompile.SetContractDeployerAllowListStatus(state, adminAddr, precompile.AllowListAdmin)
+			},
+			expectedRes: common.Hash(precompile.AllowListNoRole).Bytes(),
+			assertState: func(t *testing.T, state *state.StateDB) {
+				res := precompile.GetContractDeployerAllowListStatus(state, adminAddr)
+				assert.Equal(t, precompile.AllowListAdmin, res)
+			},
+		},
+		"read allow list with readOnly enabled": {
+			caller:         adminAddr,
+			precompileAddr: precompile.ContractDeployerAllowListAddress,
+			input: func() []byte {
+				return precompile.PackReadAllowList(noRoleAddr)
+			},
+			suppliedGas: precompile.ModifyAllowListGasCost,
+			readOnly:    true,
+			setupState: func(state *state.StateDB) {
+				precompile.SetContractDeployerAllowListStatus(state, adminAddr, precompile.AllowListAdmin)
+			},
+			expectedRes: common.Hash(precompile.AllowListNoRole).Bytes(),
+			assertState: func(t *testing.T, state *state.StateDB) {
+				res := precompile.GetContractDeployerAllowListStatus(state, adminAddr)
+				assert.Equal(t, precompile.AllowListAdmin, res)
+			},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			db := rawdb.NewMemoryDatabase()
+			state, err := state.New(common.Hash{}, state.NewDatabase(db), nil)
+			if err != nil {
+				t.Fatal(err)
+			}
+			test.setupState(state)
+
+			ret, _, err := precompile.ContractDeployerAllowListPrecompile.Run(&mockAccessibleState{state: state}, test.caller, test.precompileAddr, test.input(), test.suppliedGas, test.readOnly)
```

Comment:
nit: enforce the invariant that the remaining gas is zero (ie. we supply the exact correct amount of gas) or there is an error.